File: src/decryptage_opti.c
#include "_couches_opti.h"

/*
Fonction de déchiffrement étant donnée un chiffré, une liste de sous clé
ainsi que l'inverse des boite S et P.

L'algo présenté dans l'énoncé a été bêtement appliqué dans le sens inverse
avec les boite S et boite P inversée.
*/
int dechiffrement(int cryp, int *gencle, int* invbox_s, int* invbox_p){
    cryp = cryp ^ gencle[10];   
    for (int i = 9; i >= 0; i--){
        cryp = permut(cryp, invbox_p);
        cryp = substit(cryp, invbox_s);
        cryp = cryp ^ gencle[i];
    }
    return (cryp);
}

File: old_src_python/decryptage_opti.py
from couches_opti import permut, substit


'''
Fonction de déchiffrement. Inverse de l'algorithme present24 vu dans l'énoncé.
'''
def dechiffrement(cryp, gencle, invbox_s, invbox_p):
    etat = cryp ^ gencle[-1]
    for souscle in reversed(gencle[0:10]):
        etat = permut(etat, invbox_p)
        etat = substit(etat, invbox_s)
        etat = etat ^ souscle
    m = etat
    return m
